assigning range of values to a row in 2d numpy array

**Describe the bug**
The problem here happens when you try to assign a range of values using `np.arange` to a row of a matrix.

**To Reproduce**
Provide code to reproduce the behavior:
```python
def func(n: int, p: int):
    arr = np.zeros((n, p))
    for e in range(n):
        arr[e] = np.arange(e, e+p)
```
Provide the generated code, or the error message:
```c
/*........................................*/
void func(int64_t n, int64_t p)
{
    t_ndarray arr = {.shape = NULL};
    int64_t e;
    int64_t Dummy_0000;
    int64_t Dummy_0001;
    arr = array_create(2, (int64_t[]){n, p}, nd_double, false);
    array_fill((double)0.0, arr);
    for (e = INT64_C(0); e < n; e += INT64_C(1))
    {
        Dummy_0001 = INT64_C(0);
        for(Dummy_0000 = e; Dummy_0000 < e + p; Dummy_0000 += INT64_C(1))
        {
            arr[e, : ].nd_int64[Dummy_0001] = Dummy_0000;
            Dummy_0001 += INT64_C(1);

        }}
        free_array(arr);
    }
    /*........................................*/
```
**Error**
```bash
RuntimeError: Failed to build module
/home/nouredine.ouhaddou/lab/__pyccel__/toto.c: In function ‘func’:
/home/nouredine.ouhaddou/lab/__pyccel__/toto.c:21:20: error: expected expression before ‘:’ token
   21 |             arr[e, : ].nd_int64[Dummy_0001] = Dummy_0000;
      |    
```
**Language**
C Language

--- a/pyccel/ast/numpyext.py
+++ b/pyccel/ast/numpyext.py
@@ -36,7 +36,7 @@ from .literals       import LiteralInteger, LiteralFloat, LiteralComplex, Litera
 from .literals       import LiteralTrue, LiteralFalse
 from .literals       import Nil
 from .mathext        import MathCeil
-from .operators      import broadcast, PyccelMinus, PyccelDiv
+from .operators      import broadcast, PyccelMinus, PyccelDiv, PyccelMul, PyccelAdd
 from .variable       import (Variable, Constant, HomogeneousTupleVariable)
 
 errors = Errors()
@@ -531,6 +531,9 @@ class NumpyArange(NumpyNewArray):
     def step(self):
         return self._step
 
+    def __getitem__(self, index):
+        step = PyccelMul(index, self.step, simplify=True)
+        return PyccelAdd(self.start, step, simplify=True)
 
 #==============================================================================
 class NumpySum(PyccelInternalFunction):

--- a/pyccel/ast/utilities.py
+++ b/pyccel/ast/utilities.py
@@ -309,6 +309,9 @@ def insert_index(expr, pos, index_var):
         return type(expr)(insert_index(expr.args[0], pos, index_var),
                           insert_index(expr.args[1], pos, index_var))
 
+    elif hasattr(expr, '__getitem__'):
+        return expr[index_var]
+
     else:
         raise NotImplementedError("Expansion not implemented for type : {}".format(type(expr)))
 
@@ -356,8 +359,8 @@ def collect_loops(block, indices, new_index, language_has_vectors = False, resul
     current_level = 0
     array_creator_types = (Allocate, PythonList, PythonTuple, Concatenate, Duplicate)
     is_function_call = lambda f: ((isinstance(f, FunctionCall) and not f.funcdef.is_elemental)
-                                or (isinstance(f, PyccelInternalFunction) and not f.is_elemental
-                                    and not isinstance(f, NumpyTranspose)))
+                                or (isinstance(f, PyccelInternalFunction) and not f.is_elemental and not hasattr(f, '__getitem__')
+                                    and not isinstance(f, (NumpyTranspose))))
     for line in block:
 
         if (isinstance(line, Assign) and
@@ -400,6 +403,7 @@ def collect_loops(block, indices, new_index, language_has_vectors = False, resul
             transposed_vars = [v for v in notable_nodes if isinstance(v, NumpyTranspose)] \
                                 + [v for f in elemental_func_calls \
                                      for v in f.get_attribute_nodes(NumpyTranspose)]
+            indexed_funcs = [v for v in notable_nodes if isinstance(v, PyccelInternalFunction) and hasattr(v, '__getitem__')]
 
             is_checks = [n for n in notable_nodes if isinstance(n, PyccelIs)]
 
@@ -416,7 +420,7 @@ def collect_loops(block, indices, new_index, language_has_vectors = False, resul
             funcs           = [f for f in notable_nodes+transposed_vars if (isinstance(f, FunctionCall) \
                                                             and not f.funcdef.is_elemental)]
             internal_funcs  = [f for f in notable_nodes+transposed_vars if (isinstance(f, PyccelInternalFunction) \
-                                                            and not f.is_elemental) \
+                                                            and not f.is_elemental and not hasattr(f, '__getitem__')) \
                                                             and not isinstance(f, NumpyTranspose)]
 
             # Collect all variables for which values other than the value indexed in the loop are important
@@ -457,7 +461,7 @@ def collect_loops(block, indices, new_index, language_has_vectors = False, resul
             rank = line.lhs.rank
             shape = line.lhs.shape
             new_vars = variables
-            new_vars_t = transposed_vars
+            handled_funcs = transposed_vars + indexed_funcs
             # Loop over indexes, inserting until the expression can be evaluated
             # in the desired language
             new_level = 0
@@ -468,8 +472,8 @@ def collect_loops(block, indices, new_index, language_has_vectors = False, resul
                     indices.append(new_index('int','i'))
                 index_var = indices[rank+index]
                 new_vars = [insert_index(v, index, index_var) for v in new_vars]
-                new_vars_t = [insert_index(v, index, index_var) for v in new_vars_t]
-                if compatible_operation(*new_vars, *new_vars_t, language_has_vectors = language_has_vectors):
+                handled_funcs = [insert_index(v, index, index_var) for v in handled_funcs]
+                if compatible_operation(*new_vars, *handled_funcs, language_has_vectors = language_has_vectors):
                     break
 
             # TODO [NH]: get all indices when adding axis argument to linspace function
@@ -477,10 +481,12 @@ def collect_loops(block, indices, new_index, language_has_vectors = False, resul
                 line.rhs.ind = indices[0]
 
             # Replace variable expressions with Indexed versions
-            line.substitute(variables, new_vars, excluded_nodes = (FunctionCall, PyccelInternalFunction))
-            line.substitute(transposed_vars, new_vars_t, excluded_nodes = (FunctionCall))
+            line.substitute(variables, new_vars,
+                    excluded_nodes = (FunctionCall, PyccelInternalFunction))
+            line.substitute(transposed_vars + indexed_funcs, handled_funcs,
+                    excluded_nodes = (FunctionCall))
             _ = [f.substitute(variables, new_vars) for f in elemental_func_calls]
-            _ = [f.substitute(transposed_vars, new_vars_t) for f in elemental_func_calls]
+            _ = [f.substitute(transposed_vars + indexed_funcs, handled_funcs) for f in elemental_func_calls]
 
             # Recurse through result tree to save line with lines which need
             # the same set of for loops

--- a/pyccel/codegen/printing/ccode.py
+++ b/pyccel/codegen/printing/ccode.py
@@ -401,43 +401,6 @@ class CCodePrinter(CodePrinter):
         self.add_import(c_imports['ndarrays'])
         return buffer_array, array_init
 
-    def fill_NumpyArange(self, expr, lhs):
-        """ print the assignment of a NumpyArange
-        parameters
-        ----------
-            expr : NumpyArange
-                The node holding NumpyArange
-            lhs : Variable
-                 The left hand of Assign
-        Return
-        ------
-            String
-                Return string that contains the Assign code and the For loop
-                responsible for filling the array values
-        """
-        start  = self._print(expr.start)
-        stop   = self._print(expr.stop)
-        step   = self._print(expr.step)
-        dtype  = self.find_in_ndarray_type_registry(self._print(expr.dtype), expr.precision)
-
-        target = self.scope.get_temporary_variable(expr.dtype)
-        index  = self.scope.get_temporary_variable(NativeInteger())
-
-        self._additional_code += self._print(Assign(index, LiteralInteger(0)))
-
-        code = 'for({target} = {start}; {target} {op} {stop}; {target} += {step})'
-        code += '\n{{\n{lhs}.{dtype}[{index}] = {target};\n'
-        code += self._print(AugAssign(index, '+', LiteralInteger(1))) + '\n}}'
-        code = code.format(target = self._print(target),
-                            start = start,
-                            stop  = stop,
-                            op    = '<' if not isinstance(expr.step, PyccelUnarySub) else '>',
-                            step  = step,
-                            index = self._print(index),
-                            lhs   = lhs,
-                            dtype = dtype)
-        return code
-
     def _handle_inline_func_call(self, expr):
         """ Print a function call to an inline function
         """
@@ -1747,8 +1710,6 @@ class CCodePrinter(CodePrinter):
             return prefix_code+self.copy_NumpyArray_Data(expr)
         if isinstance(rhs, (NumpyFull)):
             return prefix_code+self.arrayFill(expr)
-        if isinstance(rhs, NumpyArange):
-            return prefix_code+self.fill_NumpyArange(rhs, lhs)
         lhs = self._print(expr.lhs)
         rhs = self._print(expr.rhs)
         return prefix_code+'{} = {};\n'.format(lhs, rhs)

--- a/tests/epyccel/modules/arrays.py
+++ b/tests/epyccel/modules/arrays.py
@@ -1813,6 +1813,12 @@ def arr_arange_6():
     a = np.arange(20, 1, -1.1)
     return np.shape(a)[0], a[0], a[-1]
 
+def arr_arange_7(arr : 'int[:,:]'):
+    import numpy as np
+    n, m = arr.shape
+    for i in range(n):
+        arr[i] = np.arange(i, i+m)
+
 def iterate_slice(i : int):
     import numpy as np
     a = np.arange(15)

--- a/tests/epyccel/test_arrays.py
+++ b/tests/epyccel/test_arrays.py
@@ -3998,6 +3998,17 @@ def test_numpy_arange_negative_step_2(language):
     f2 = epyccel(f1, language = language)
     np.testing.assert_array_almost_equal(f1(), f2(), decimal = 9)
 
+def test_numpy_arange_into_slice(language):
+    f1 = arrays.arr_arange_7
+    f2 = epyccel(f1, language = language)
+    n = randint(2,10)
+    m = randint(2,10)
+    x = np.array(100*np.random.random((n,m)), dtype=int)
+    x_expected = x.copy()
+    f1(x_expected)
+    f2(x)
+    np.testing.assert_array_almost_equal(x, x_expected, decimal = 9)
+
 def test_iterate_slice(language):
     f1 = arrays.iterate_slice
     f2 = epyccel(f1, language = language)

--- a/tests/epyccel/test_arrays.py
+++ b/tests/epyccel/test_arrays.py
@@ -4009,7 +4009,7 @@ def test_numpy_arange_into_slice(language):
     x_expected = x.copy()
     f1(x_expected)
     f2(x)
-    np.testing.assert_allclose(f1(), f2(), rtol=RTOL, atol=ATOL)
+    np.testing.assert_allclose(x, x_expected, rtol=RTOL, atol=ATOL)
 
 def test_iterate_slice(language):
     f1 = arrays.iterate_slice

--- a/tests/epyccel/test_arrays.py
+++ b/tests/epyccel/test_arrays.py
@@ -3967,6 +3967,8 @@ def test_arrs_2d_negative_index(language):
 #==============================================================================
 # TEST : NUMPY ARANGE
 #==============================================================================
+RTOL = 1e-12
+ATOL = 1e-16
 
 def test_numpy_arange_one_arg(language):
     f1 = arrays.arr_arange_1
@@ -3981,7 +3983,7 @@ def test_numpy_arange_two_arg(language):
 def test_numpy_arange_full_arg(language):
     f1 = arrays.arr_arange_3
     f2 = epyccel(f1, language = language)
-    np.testing.assert_array_almost_equal(f1(), f2(), decimal=9)
+    np.testing.assert_allclose(f1(), f2(), rtol=RTOL, atol=ATOL)
 
 def test_numpy_arange_with_dtype(language):
     f1 = arrays.arr_arange_4
@@ -3991,28 +3993,28 @@ def test_numpy_arange_with_dtype(language):
 def test_numpy_arange_negative_step(language):
     f1 = arrays.arr_arange_5
     f2 = epyccel(f1, language = language)
-    np.testing.assert_array_almost_equal(f1(), f2(), decimal = 9)
+    np.testing.assert_allclose(f1(), f2(), rtol=RTOL, atol=ATOL)
 
 def test_numpy_arange_negative_step_2(language):
     f1 = arrays.arr_arange_6
     f2 = epyccel(f1, language = language)
-    np.testing.assert_array_almost_equal(f1(), f2(), decimal = 9)
+    np.testing.assert_allclose(f1(), f2(), rtol=RTOL, atol=ATOL)
 
 def test_numpy_arange_into_slice(language):
     f1 = arrays.arr_arange_7
     f2 = epyccel(f1, language = language)
-    n = randint(2,10)
-    m = randint(2,10)
-    x = np.array(100*np.random.random((n,m)), dtype=int)
+    n = randint(2, 10)
+    m = randint(2, 10)
+    x = np.array(100 * np.random.random((n, m)), dtype=int)
     x_expected = x.copy()
     f1(x_expected)
     f2(x)
-    np.testing.assert_array_almost_equal(x, x_expected, decimal = 9)
+    np.testing.assert_allclose(f1(), f2(), rtol=RTOL, atol=ATOL)
 
 def test_iterate_slice(language):
     f1 = arrays.iterate_slice
     f2 = epyccel(f1, language = language)
-    i = randint(2,10)
+    i = randint(2, 10)
     assert f1(i) == f2(i)
 
 ##==============================================================================

